Re-diffed for 2.6.14.7 by Kerin Millar <kerframil@gmail.com>:

From stable-bounces@linux.kernel.org  Wed Feb  1 21:41:00 2006
Date: Thu, 02 Feb 2006 00:09:05 -0500
From: Dmitry Torokhov <dtor_core@ameritech.net>
To: stable@kernel.org
Cc: 
Subject: Input: iforce - fix detection of USB devices

Recent conversion to wait_event_interruptible_timeout() caused
USB detection routine erroneously report timeouts for perfectly
working devices.

Signed-off-by: Dmitry Torokhov <dtor@mail.ru>
Signed-off-by: Chris Wright <chrisw@sous-sol.org>
---

--- linux-2.6.14.7.orig/drivers/input/joystick/iforce/iforce-packets.c	2005-10-27 19:02:08.000000000 -0500
+++ linux-2.6.14.7/drivers/input/joystick/iforce/iforce-packets.c	2006-02-06 21:42:22.000000000 -0500
@@ -166,9 +166,9 @@ void iforce_process_packet(struct iforce
 		iforce->expect_packet = 0;
 		iforce->ecmd = cmd;
 		memcpy(iforce->edata, data, IFORCE_MAX_LENGTH);
-		wake_up(&iforce->wait);
 	}
 #endif
+	wake_up(&iforce->wait);
 
 	if (!iforce->type) {
 		being_used--;
@@ -263,7 +263,7 @@ int iforce_get_id_packet(struct iforce *
 		wait_event_interruptible_timeout(iforce->wait,
 			iforce->ctrl->status != -EINPROGRESS, HZ);
 
-		if (iforce->ctrl->status != -EINPROGRESS) {
+		if (iforce->ctrl->status) {
 			usb_unlink_urb(iforce->ctrl);
 			return -1;
 		}
--- linux-2.6.14.7.orig/drivers/input/joystick/iforce/iforce-usb.c	2005-10-27 19:02:08.000000000 -0500
+++ linux-2.6.14.7/drivers/input/joystick/iforce/iforce-usb.c	2006-02-06 21:42:55.000000000 -0500
@@ -95,7 +95,6 @@ static void iforce_usb_irq(struct urb *u
 		goto exit;
 	}
 
-	wake_up(&iforce->wait);
 	iforce_process_packet(iforce,
 		(iforce->data[0] << 8) | (urb->actual_length - 1), iforce->data + 1, regs);
 
